### PR TITLE
Fix #3607 - constant string class reference with leading backslash

### LINKED ISSUE #3607: Unused class/method misses const strings with leading backslash
This one is fine:
https://psalm.dev/r/17dc5a2e45

Psalm misses this one:
https://psalm.dev/r/c5d8cd84d8

But adding a leading backslash is legal:
https://3v4l.org/R1B3f

Thanks!

### PR DESCRIPTION
**Pull request by iluuu1994**
Let me know if this fix is appropriate 🙂

### CODE REVIEW

**Comment by iluuu1994** [comment]
Hm, this breaks something... I'll take a look.

**Comment by iluuu1994** [comment]
@muglug Very peculiar. As soon as I touch `$fq_class_name` in `MethodIdentifier::__construct` Psalm breaks... Even just fetching it and then doing nothing.

```php
$fq_class_name;
```

Any idea what's going on here?

File: src/Psalm/Internal/MethodIdentifier.php [hunk 1]
```diff
@@ -13,6 +13,9 @@ class MethodIdentifier
      */
     public function __construct(string $fq_class_name, string $method_name)
     {
+        // remove trailing backslash if it exists
```
Comment (staabm) [inline src/Psalm/Internal/MethodIdentifier.php:16]: The comment is wrong... it removes a leading backslash, no?

Comment (staabm) [reply src/Psalm/Internal/MethodIdentifier.php:16]: Could this just be a `ltrim(...)` instead?

Comment (iluuu1994) [reply src/Psalm/Internal/MethodIdentifier.php:16]: > The comment is wrong... it removes a leading backslash, no?

Oh yeah, I just moved that code block, the comment was already wrong :) 

> Could this just be a ltrim(...) instead to ease reading...?

That would remove multiple backslashes which would make Psalm thing that `\\Foo` is valid but it's not.

**Comment by muglug** [comment]
> @muglug Very peculiar. As soon as I touch $fq_class_name in MethodIdentifier::__construct Psalm breaks... Even just fetching it and then doing nothing... Any idea what's going on here?

Yeah, I didn't bother to define property types because Psalm was able to infer them in the scanning stage. Adding the input sanitisation breaks that (though you can always add the property types yourself).

I'd encourage you instead to do the sanitisation before the call to `new MethodIdentifier` – that way you're not penalising every call (which almost never comes from a specific string).

**Comment by iluuu1994** [comment]
> I'd encourage you instead to do the sanitisation before the call to new MethodIdentifier – that way you're not penalising every call (which almost never comes from a specific string).

Ok, I'll do that 👍

**Comment by iluuu1994** [comment]
@muglug Done 🙂

**Comment by muglug** [comment]
Thanks!

**Pull request merged**